pennylane/optimize/rotoselect.py

```diff
@@ -0,0 +1,178 @@
+    element-wise arc tangent of :math:`x/y` choosing the quadrant correctly, avoiding, in
+    particular, division-by-zero when :math:`y = 0`.
+
+    The algorithm is described in further detail in `Ostaszewski et al. (2019) <https://arxiv.org/abs/1905.09692>`_.
+
+    Keyword Args:
+        possible_generators (list[~.Operation]): List containing the possible ``pennylane.ops.qubit`` operators
+            that are allowed in the circuit. Default is the set of Pauli matrices :math:`\{X, Y, Z\}`.
+
+    **Example:**
+
+    Initialize the optimizer, set the initial values of ``x``, choose the initial generators to
+    start with, and set the number of steps to optimize over.
+
+    >>> opt = qml.optimize.RotoselectOptimizer()
+    >>> x = [0.3, 0.7]
+    >>> generators = [qml.RX, qml.RY]
+    >>> n_steps = 1000
+
+    Define a cost function based on a PennyLane circuit
+
+    >>> def cost(x, generators):
+    >>>     X_1, Y_2 = circuit(x, generators=generators)
+    >>>     return 0.2 * X_1 + 0.5 * Y_2
+
+    Run the optimization step-by-step for ``n_steps`` steps.
+
+    >>> cost_rotosel = []
+    >>> for _ in range(n_steps):
+    >>>     cost_rotosel.append(cost(x, generators))
+    >>>     x, generators = opt.step(cost, x, generators)
+
+    The optimized values for x should now be stored in ``x`` together with the optimal gates for
+    the circuit, while steps-vs-cost can be seen by plotting ``cost_rotosel``.
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, **kwargs):
+        self.possible_generators = kwargs.get("possible_generators", [qml.RX, qml.RY, qml.RZ])
+
+    def step(self, objective_fn, x, generators):
+        """Update x with one step of the optimizer.
+
+        Args:
+            objective_fn (function): The objective function for optimization. It should take an
+                array of the values ``x`` and a list of the gates ``generators`` as inputs, and
+                return a single value.
+            x (Union[array[float], float]): NumPy array containing the initial values of the
+                variables to be optimized over, or a single float with the initial value.
+            generators (list[~.Operation]): List containing the initial ``pennylane.ops.qubit``
+                operators to be used in the circuit and optimized over.
+
+        Returns:
+            array: The new variable values :math:`x^{(t+1)}` as well as the new generators.
+        """
+        # make sure that x is an array of shape (1, -1)
+        if np.ndim(x) == 0:
+            x = np.array([x], dtype=float)
+        try:
+            x = np.array(x, dtype=float)
+            assert np.ndim(x) == 1
+        except (AssertionError, ValueError):
+            raise ValueError(
+                "Input must be either an array with dimension 1 or a float, and not x = {}.".format(
+                    x
+                )
+            )
```

Comment:
What is the reason for the shape validation here? The other optimizers make use of the `_flatten()` and `unflatten()` functions in `pennylane.utils`. These are quite useful, in that they allow _any_ arbitrarily nested sequence to be flattened to a 1D numpy array. Once the parameter update step is complete, the updated 1D parameter list can be 'unflattened' back to the original input shape.

Reply:
(and going back to Antal's comment, this allows for the type of parameter `x` to simply be `Sequence[float] or float`, rather than restricting to `array` or `list` 🙂 )

Reply:
Updated both optimizers to use ``_flatten`` and ``unflatten``. Looks much better now. Thanks!

Reply:
Also updated to ``Sequence``. 🙂